### PR TITLE
Get components of a GCP path

### PR DESCRIPTION
**Pull request by EddieLF**
Finally figured out the regex pattern to match the various components of a gcp path for almost all cpg buckets, no thanks to ChatGPT (seriously, it seems to suck at regex). 

[See here for some tests](https://regex101.com/r/r8ynHJ/1).

### CODE REVIEW

File: cpg_utils/cloud.py [hunk 1]
```diff
@@ -322,3 +323,42 @@ def is_member_in_cached_group(
...
+    subdirectory, and the file name.
+    """
+
+    bucket_types = ['archive', 'hail', 'main', 'test', 'release']
+
+    rmatch_str = (
+        r'gs://(?P<bucket>cpg-(?P<dataset>[\w-]+)-(?P<bucket_type>['
+        + '|'.join(s for s in bucket_types)
+        + r']+[-\w]*))/(?P<suffix>.+/)?(?P<file>.*)$'
+    )
```
Comment (MattWellie) [inline cpg_utils/cloud.py]: this is cool, but unless it's parametrised from config you may as well just make the fixed string here?

File: cpg_utils/cloud.py [hunk 2]
```diff
@@ -322,3 +323,42 @@ def is_member_in_cached_group(
...
+
+        return {
+            'bucket_name': bucket_name,
+            'dataset': dataset,
+            'bucket_type': bucket_type,
+            'subdir': subdir,
+            'file': file,
+        }
+
+    return None
```
Comment (MattWellie) [inline cpg_utils/cloud.py]: I'd be happy to either return contents or fail here. It's a very specific method, and being able to return `None` in what is probably an error state just makes the output a bit harder to predict elsewhere

```suggestion

    if path_match := gspath_pattern.match(path):
        path_components = path_match.groups()

        bucket_name = path_components[0]
        dataset = path_components[1]
        bucket_type = path_components[2]
        subdir = path_components[3]
        file = path_components[4]

        return {
            'bucket_name': bucket_name,
            'dataset': dataset,
            'bucket_type': bucket_type,
            'subdir': subdir,
            'file': file,
        }

    raise ValueError('The input String did not match a valid GCP path')

```

File: cpg_utils/cloud.py [hunk 3]
```diff
@@ -322,3 +323,42 @@ def is_member_in_cached_group(
...
+        subdir = path_components[3]
+        file = path_components[4]
+
+        return {
+            'bucket_name': bucket_name,
+            'dataset': dataset,
+            'bucket_type': bucket_type,
+            'subdir': subdir,
+            'file': file,
+        }
```
Comment (MattWellie) [inline cpg_utils/cloud.py]: I just discovered this, neat shortcut 

```suggestion
        return path_match.groupdict()
```

**Review by MattWellie — CHANGES_REQUESTED** [review]
I think a failure to match should be an error, not a None. The other suggestions are syntax changes but the functionality is the same

**Comment by MattWellie** [comment]
Also all the suggestions I added overlap in a really stupid way
...

```python
def get_path_components_from_gcp_path(path: str) -> dict[str, str]:
    """
    Return the {bucket_name}, {dataset}, {bucket_type}, {subdir}, and {file} for GS only paths
    Uses regex to match the full bucket name, dataset name, bucket type (e.g. 'test', 'main-upload', 'release'),
    subdirectory, and the file name.
    """

    bucket_types = ['archive', 'hail', 'main', 'test', 'release']

    # compile pattern matching all CPG bucket formats
    gspath_pattern = re.compile(
        r'gs://(?P<bucket>cpg-(?P<dataset>[\w-]+)-(?P<bucket_type>['
        + '|'.join(s for s in bucket_types)
        + r']+[-\w]*))/(?P<suffix>.+/)?(?P<file>.*)$'
    )

    # if a match succeeds, return the key: value dictionary
    if path_match := gspath_pattern.match(path):
        return path_match.groupdict()

    # raise an error if the input String was not a valid CPG bucket path
    raise ValueError('The input String did not match a valid GCP path')
```

File: cpg_utils/cloud.py (hunk 1 above)
Comment (EddieLF) [reply cpg_utils/cloud.py]: Not really sure what you mean by this, are you saying to scrap the bucket_types list and add them into the string?

File: cpg_utils/cloud.py (hunk 3 above)
Comment (EddieLF) [reply cpg_utils/cloud.py]: Now this I love

File: cpg_utils/cloud.py (hunk 1 above)
Comment (MattWellie) [reply cpg_utils/cloud.py]: Pretty much. I actually like keeping this in, as it shows the syntax if we want to entertain a wider range of bucket categories later.

**Comment by lgruen** [comment]
Have you tried https://regex.ai/?

**Comment by EddieLF** [comment]
> Have you tried https://regex.ai/?

You keep giving me amazing tools to bookmark @lgruen! This is another awesome one, alongside phind.com

**Review by MattWellie — APPROVED** [review]

**Pull request merged**